blog: summarize achievements of revival milestone 2 and plans for milestone 3 (#228)
MovingBlocks/movingblocks.github.com: blog/2023-11-20-second-revival-milestone-unveiled/index.md

---
posttype: blog
title: "Revival Milestone 2 Unveiled - Elevating Quality, Boosting Performance, Advancing to Java 17!"
cover: "./cover.jpg"
description: "Unveiling Terasology Revival Milestone 2 - where quality meets performance, and Java 17 takes the spotlight. We’re celebrating our achievements and planning our path ahead. Ready to join the journey? New contributors welcome!"
author: "Niruandaleth"
date: "2023-11-20"
tags: ["Announcement"]
---

> _Unveiling Terasology Revival Milestone 2 - where quality meets performance, and Java 17 takes the spotlight.

Reviewer: We have been working in public the whole time, so this is not an "unveiling" as such. I think "Announcing" is more accurate here.

Author: fair point 👍

We’re celebrating our achievements and planning our path ahead.

> Ready to join the journey? New contributors welcome!_

Reviewer: I think a call to action is a bit premature here. Maybe a bit later in the post.

Author: that's the blog post description / summary, so I'll leave it in 🙂

Reviewer: I did not realise that was how the syntax worked. That should be fine there then.

Hey there, fellow gooeys! Welcome to Terasology, where blocks meet potential ✨

> Our project was born in 2011 in the desire to provide a modular canvas for diverse gameplay ideas in a dynamic voxel world with creative possibilities as endless as the blocks themselves.

Reviewer: ```suggestion
Our project was born in 2011 with the desire to provide a modular canvas for diverse gameplay ideas in a dynamic voxel world with creative possibilities as endless as the blocks themselves.
```

Reviewer: Although I'm not sure that we can confidently assert this. The modularity ideas may have come later-on.

Author: I'm pretty sure that modularity was a goal from the start, so I'll leave that in (but adopting the suggestion).

> While we wear our history with pride, our codebase outgrew our maintainer capacity over the years, collecting rust and dust in the face of dwindling contributor and contribution numbers.

Reviewer: > While we wear our history with pride,

Do we? The evolution of the codebase is interesting to be sure but this is about looking forward, not back. I think that bit at the beginning can probably be removed.

Author: I do think we (although maybe not all) do. I myself am not happy with everything that happened in the project's history, both before and since I joined, but I am proud that the project survived that long and I'm proud of everybody that contributed (in whatever way, extent, and quality).
I agree that this is about looking forward, not back, but it is also about our revival effort and a reader might not have the context why there is a revival in the first place, so I'd like to leave it in to provide that context.

> Though small, our gooey family is welcoming and inclusive, with room for all voices. Each line of code, every idea, and every contribution counts. 

Reviewer: This line feel redundant. Saying explicitly that we're welcoming and inclusive may give the opposite impression (people value deeds, not words). I also wonder if it's even possible to be both welcoming and uninclusive. See also the family comments below (it is a bit cliché).

Author: Agree to disagree here (except for the family part). A reader doesn't necessarily know our community and having heard from other similar projects about not-so-nice communities, I think this is really something we can be proud of and mention. And yes, I think you can be welcoming and uninclusive depending on whom you welcome (and whom you may not or in a different way). I'll leave this in.

Recently, our shared love for Terasology propelled us into an exciting phase of revival, where we not only seek to breathe new life into the project but also to invigorate our community.

> This is not just about coding — it's about all of us coming together to unlock Terasology's true potential and build a voxel game that we enjoy.

Reviewer: I think the point you're trying to make here is that we're not just doing codebase restructuring, refactoring or generic overhauls. We're out to make the game better, on all fronts. It is about coding, although eventually we are going to need more than that too.

Author: Yes and no ... or rather ... not only.
For me (although that might be only me, don't know), it is about building something together, not only a game on a technical (including UI, UX, docs) level, but also on a community level. Do you know what I mean?

Also, was that a comment suggesting changing the wording or just a statement of thought?

Reviewer: My comment was more a reaction to this part:
> This is not just about coding

Coming back to it, this is actually fine. What I understood of it is clear enough from the text already present.

The focus right now? Modernizing and giving our codebase a facelift. We believe that the foundation of our project should be as robust and adaptable as our creative aspirations 💡

> The first milestone brought us new contributors, upgraded dependencies, and first progress towards a trimmer, sleeker codebase.

Reviewer: > first progress towards a trimmer, sleeker codebase

At least in my vocabulary, a trimmer is a sharp object. To trim something is to cut something short. Did you possibly mean "slimmer", "thinner" or "leaner"?

Author: I don't recall anymore how I ended up with "trimmer" tbh 😅
But yeah, I wanted to go for the direction of slimmer / thinner / leaner 👍

Now it's with pride that we announce the completion of our second revival milestone 🎉
And that's not the end - the third revival milestone already unfurls its sails, and we invite you to join us on our voyage into the future of Terasology ⛵

Let's explore how our second revival milestone added to breathing new life into the codebase, setting the stage for Terasology's grand comeback!

## 📜 Definition of Done: A Checklist of Progress

One of our major achievements is the seamless transition to Java 17 driven by _@BSA_, laying the groundwork for a more robust and modern Terasology engine 🛠️
The upgrade not only marks a technical advancement but also signifies our commitment to moving Terasology towards the cutting edge of technology.
If you're already contributor, please refer to https://github.com/MovingBlocks/Terasology/issues/5177 for guidance on what this means for your Terasology workspace.
For new contributors, our updated [Contributor Quickstart Guide](https://terasology.org/Terasology/#/Contributor-Quick-Start) points to the now required Java 17 JDK 💡

> Targeting a more user-friendly interface for customizing Terasology worlds in the advanced game setup, _@Niruandaleth_ merged the universe setup screen and world pregeneration screen into a more streamlined flow. With major support from recently joined _@soloturn_, she also pursued quality improvements. Checkstyle rule configuration has been optimized to reduce noise and various warnings flagged by different QA tools like checkstyle, spotbugs, and pmd have been diligently addressed, adding to a cleaner, more efficient codebase.

Reviewer: > With major support from recently joined _soloturn_, she also pursued quality improvements.

The "she" here is a bit ambigous. On a second reading it becomes clear that this is likely `Niruandaleth` but this sentence could be re-worded a bit.

Author: I see your point but I'm struggling how to re-word that best 🤔 any suggestions @BenjaminAmos ?

Reviewer: Here's a couple of ideas:

```suggestion
Targeting a more user-friendly interface for customizing Terasology worlds in the advanced game setup, _@Niruandaleth_ merged the universe setup screen and world pregeneration screen into a more streamlined flow. With major support from recently joined _@soloturn_, together they also pursued quality improvements. Checkstyle rule configuration has been optimized to reduce noise and various warnings flagged by different QA tools like checkstyle, spotbugs, and pmd have been diligently addressed, adding to a cleaner, more efficient codebase.
```

```suggestion
Targeting a more user-friendly interface for customizing Terasology worlds in the advanced game setup, _@Niruandaleth_ merged the universe setup screen and world pregeneration screen into a more streamlined flow. Along with major support from recently joined _@soloturn_, quality improvements were pursued. Checkstyle rule configuration has been optimized to reduce noise and various warnings flagged by different QA tools like checkstyle, spotbugs, and pmd have been diligently addressed, adding to a cleaner, more efficient codebase.
```

```suggestion
Targeting a more user-friendly interface for customizing Terasology worlds in the advanced game setup, _@Niruandaleth_ merged the universe setup screen and world pregeneration screen into a more streamlined flow. Recently joined _@soloturn_ worked with _@Niruandaleth_ to pursue quality improvements. Checkstyle rule configuration has been optimized to reduce noise and various warnings flagged by different QA tools like checkstyle, spotbugs, and pmd have been diligently addressed, adding to a cleaner, more efficient codebase.
```

```suggestion
Targeting a more user-friendly interface for customizing Terasology worlds in the advanced game setup, _@Niruandaleth_ merged the universe setup screen and world pregeneration screen into a more streamlined flow. _@Niruandaleth_, together with major support from recently joined _@soloturn_, also pursued quality improvements. Checkstyle rule configuration has been optimized to reduce noise and various warnings flagged by different QA tools like checkstyle, spotbugs, and pmd have been diligently addressed, adding to a cleaner, more efficient codebase.
```

Author: Thanks for the suggestions, I played around with a few of them and ended up at the final
> Recently joined _@soloturn_ provided major support in pursuing various quality improvements.

Quality improvements will continue in our next revival milestone - feel free to come aboard! ⛵

The quest for optimal performance is a journey, not a destination 🚀
Milestone 2 marks the beginning of our performance investigations, paving the way for future optimizations that will elevate the gaming experience for our users.
A first success was _@skaldarnar_ identifying leftovers of the VR integration removed in revival milestone 1.
Removing them not only further decluttered our codebase but contributed to a more resource-efficient gameplay experience by halving the required rendering passes.

Together, our community discussed the benefit of having documentation closer to our code.

> As a result, we integrated our Terasology engine wiki into our codebase, making it our new ["Knowledge Base"](https://terasology.org/Terasology/#/) provided using Docsify on GitHub Pages, allowing to co-locate code and associated documentation changes in commits and PRs.

Reviewer: > allowing to co-locate code and associated documentation changes in commits and PRs

This is a very developer-specific way of putting it. The goal is to maintain code and documentation changes in sync. The fact that this is done in the same change units (commits) is irrelevant to players.

Author: fair point, I'll adjust that 👍

> All links to our old wiki should now point to our ["Knowledge Base"](https://terasology.org/Terasology/#/). If you find any broken links, please report them on Discord or using a GitHub issue 💡

Reviewer: You might want to mention that we are unable to do anything about external links to https://github.com/MovingBlocks/Terasology/wiki/. Those will have stopped working when the wiki tab was disabled.

Author: "External" as in some third-party reference outside of our control?
We can not fix it alone, but we could approach whoever maintains the thing that references our wiki, right?

Reviewer: We can approach other places that referenced the wiki if we choose to. Your understanding of "External" is the same as mine.

Essentially, whilst we've tried to update all the pages we control, if you see a link to https://github.com/MovingBlocks/Terasology/wiki/ then you'll need to go to https://terasology.org/Terasology/#/ instead. That's the jist of what we might want to convey. On the other hand, this may not be a significant issue (all old wiki links should redirect to the project README anyway, which contains a link to the knowledge base).

Author: I rephrased it to
> If you find any links to our old wiki that we might have missed, please report them on Discord or using a GitHub issue 💡

Maybe that makes the call to action if anybody notices a broken wiki link clearer.

Next to these bigger efforts, a lot of smaller discussions and activities happened across our GitHub repositories and Discord channels.
_@BSA_, _@Niruandaleth_, and _@skaldarnar_ met once per week virtually to discuss milestone progress, agree on strategic questions, and collaborate on blockers and reviews.
Over the course of our second revival milestone, _@Cervator_ provided guidance for new community members and revivers alike, sharing his knowledge of the project history and in particular its infrastructure.
Long-time contributor _@rzats_ visited the project, bringing Ukrainian locale up to date.

> _@qwc_ will be hosting the Valheim server again for our second revival milestone completion celebrations, adding a touch of community spirit as the year slowly comes to its end.

Reviewer: > _qwc_ will be hosting the Valheim server again for our second revival milestone completion celebrations, adding a touch of community spirit as the year slowly comes to its end.

1. This has already happened.
2. I don't think players need to know about this part. When Terasology multiplayer is in a playable state again then maybe we can mention a playtest but another game entirely may give the wrong impression.

Author: (1) it hadn't when I wrote this 😅 but as it's more than a month later now, I should indeed change that
(2) I'd like to keep that in as I think it shows that we do something fun together as a community even though Terasology is not yet in the shape to use it. and even though Valheim is another game entirely, it is in many ways very close to Terasology and not only a fun game to play but a good source of inspiration for a level of playability I hope Terasology some day will reach at least to some degree.

> To each contributor, thank you for your dedication, passion, and relentless pursuit of improving Terasology!

Reviewer: > To each contributor

Or maybe "To all our contributors"? Either would work but one is more individualistic and the other more collective.

Author: I do like "To all our contributors" better here, yes, thanks for the suggestion!

> Milestone 2 is not just a checkpoint; it's a testament to what we can achieve together.

Reviewer: >  Milestone 2 is not just a checkpoint; it's a testament to what we can achieve together.

This is slightly downplaying the the work in a subtle way. Rather then saying what it's not, talk about what it is.

Reviewer: ```suggestion
Milestone 2 is more than just a checkpoint; it's a testament to what we can achieve together.
```

I feel that this variant is slightly more positive.

Author: Oh yes, that variant absolutely is more positive, I like it 👍

> Here's to our community, our contributors, and our potential 🌈

Reviewer: ```suggestion
Here's to our community, our contributors, and our potential!
```

This feels like it deserves an exclamation mark.

Author: Absolutely! 👍

> 🎉 In total, we merged [27 engine PRs](https://github.com/MovingBlocks/Terasology/pulls?page=1&q=is%3Apr+sort%3Aupdated-desc+is%3Amerged+merged%3A2023-09-11..2023-11-19) and [20 PRs on libraries and side projects](https://github.com/search?q=-repo%3Amovingblocks%2Fterasology+org%3Amovingblocks+is%3Apr+merged%3A2023-09-11..2023-11-19+&type=pullrequests&p=2) such as gestalt, the launcher, or the module site.

Reviewer: Maybe this sentence should be at the beginning of the milestone summary?

Author: In the TeraSaturdays I think we always had it at the end of the summary, but I can also put it at the beginning 🤷 I don't have a particular preference there.

## 🚀 Terasology Revival Milestone 3: New Contributors Welcome

> Embarking on the journey of our third revival milestone, our repositories are wide open for new contributions by friendly people joining our gooey family! 🌟

Reviewer: > our repositories are wide open for new contributions by friendly people joining our gooey family

Family references are often misunderstood. We are a group of like-minded individuals working towards a common goal. This is very much a cultural thing but I personally would percieve the description of our community as a family to be rather unprofessional. At least for me, family, friends and collegues are all very different things with very different levels of involvement. Referring to us as a "community" is the most appropriate term and does have the same warm positive connotations.

Author: Fair point, I'll change it.

I get what you mean, but I think for me personally, I do have different "kinds of families". I have my blood family (relatives), I have my work family (colleagues), and I have my friends who are kind of "the people I chose to be my family".
But yeah, it is a very personal and in many ways probably also a cultural thing.

Whether you're a seasoned developer or just getting started, there's a place for you in our dynamic community. Let's build, innovate, and play together! 🚀💬

> Let's delve into the tasks that lie ahead, each playing a crucial role in shaping the future of Terasology.

Reviewer: ```suggestion
Let's delve into the tasks that lie ahead.
```

A short sentence is often used as an introduction into a factual overview. This seems more natural to me without the last bit.

Author: Fine with me. I have a tendency to write too many and too long sentences 😅

> If you'd like to contribute to a specific task, feel free to ping the respective revivers on our [Discord](https://discord.gg/terasology).

Reviewer: This is a remark. I would say that it can either go in brackets (emphasising optional reading), or it can go at the bottom after summarising the tasks.

Author: I'd like to keep it at the top to allow the reader to take not of revivers associated with tasks the reader is interested in on the first read-through to avoid the need for having to go back through the text to find the respective information.
Emphasising optional reading is a good idea, although I'm a bit split on that as I want people to read and follow that remark, so kind of encouraging to not read that part by default... not sure that's what I want 😅 I'll think about that some and then decide what to do with it.

Author: I think I'll leave it where it is.
